server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1785,42 +1964,62 @@ void handleSelfCertificate(Consumer consumer, Pool pool,
      */
     private class NewHandler implements EntitlementHandler{
         @Override
-        public Entitlement handleEntitlement(Consumer consumer, Pool pool,
-            Entitlement entitlement, int quantity) {
-            Entitlement newEntitlement = new Entitlement(pool, consumer, quantity);
-            consumer.addEntitlement(newEntitlement);
-            pool.getEntitlements().add(newEntitlement);
-            return newEntitlement;
+        public Map<String, Entitlement> handleEntitlement(Consumer consumer,
+                Map<String, PoolQuantity> poolQuantities,
+                Map<String, Entitlement> entitlements) {
+            Map<String, Entitlement> result = new HashMap<String, Entitlement>();
+            for (Entry<String, PoolQuantity> entry : poolQuantities.entrySet()) {
+                Entitlement newEntitlement = new Entitlement(entry.getValue().getPool(), consumer, entry
+                        .getValue().getQuantity());
+                result.put(entry.getKey(), newEntitlement);
+                consumer.addEntitlement(newEntitlement);
```

Comment:
This is the reason why revoke doesn't work in PoolManagerFunctionalTest.testRevocation() (the "TODO fix me" that you've left there) without EntityManager.clear().  The instance newEntitlement is being added to consumer.entitlements collection before being persisted. Because we implement equals/hashcode by @Id, this is fundamentally flawed (and actually disalowed even on Java level, see note in https://docs.oracle.com/javase/8/docs/api/java/util/Set.html). The revokeEntitlements in the PoolManagerFunctionalTest.testRevocation() method will try to remove the newEntitlement (revoked one) from the collection but will fail because the entitlement will have @Id already assigned and thus equals/hashCode value changes.

@mstead @vritant Guys I will send you shortly a longer explanation of this problem, because I feel its something that we do in our code base regulary (including me) and its adding so much pain to work with Hibernate